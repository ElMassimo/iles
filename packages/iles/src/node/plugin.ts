import path from 'path'
import fs from 'fs'
import { build as viteBuild } from 'vite'
import type { Manifest, PluginOption, ResolvedConfig, ResolveFn, UserConfig } from 'vite'
import type { ViteSSGOptions } from '@mussi/vite-ssg'

import Vue from '@vitejs/plugin-vue'
import Pages from 'vite-plugin-pages'
import Layouts from 'vite-plugin-vue-layouts'
import IconsResolver from 'unplugin-icons/resolver'
import ViteComponents from 'unplugin-vue-components/vite'
import VueJSX from '@vitejs/plugin-vue-jsx'
import XDM from 'vite-plugin-xdm'

import matter from 'gray-matter'
import glob from 'fast-glob'
import chalk from 'chalk'
import createDebugger from 'debug'
import { escapeRegex, pascalCase, serialize } from './string'
import type { IslandsConfig } from './shared'
import { uniq } from './array'
import { parseImports, rebaseImports } from './parse'

import { resolveAliases, APP_PATH, DIST_CLIENT_PATH, SITE_DATA_REQUEST_PATH, ROUTES_REQUEST_PATH, USER_CONFIG_REQUEST_PATH } from './alias'

const debug = {
  mdx: createDebugger('islands:mdx'),
  wrap: createDebugger('islands:wrap'),
  resolve: createDebugger('islands:resolve'),
  build: createDebugger('islands:build'),
}

function resolveManifestEntries (manifest: Manifest, entryNames: string[]): string[] {
  return entryNames.flatMap((entryName) => {
    const entry = manifest[entryName]
    return [entry.file, ...resolveManifestEntries(manifest, entry.imports || [])]
  })
}

function stringifyPreload (manifest: Manifest, hrefs: string[]) {
  return uniq(resolveManifestEntries(manifest, hrefs))
    .map(href => `<link rel="modulepreload" href="${path.join(base, href)}" crossorigin/>`)
    .join('')
}

function parseId (id: string) {
  const index = id.indexOf('?')
  if (index < 0) {
    return { path: id, query: {} }
  }
  else {
    // @ts-ignore
    const query = Object.fromEntries(new URLSearchParams(id.slice(index)))
    return {
      path: id.slice(0, index),
      query,
    }
  }
}

function filenameFromRoute (route: string) {
  const relativeRoute = (route.endsWith('/') ? `${route}index` : route).replace(/^\//g, '')
  const filename = `${relativeRoute}.html`
  return path.join(outDir, filename)
}

function buildLog (text: string, count: number | undefined) {
  console.log(`
${chalk.gray('[islands]')} ${chalk.yellow(text)}${count ? chalk.blue(` (${count})`) : ''}`)
}

async function replaceAsync (str: string, regex: RegExp, asyncFn: (...groups: string[]) => Promise<string>) {
  const promises = Array.from(str.matchAll(regex))
    .map(([match, ...args]) => asyncFn(match, ...args))
  const replacements = await Promise.all(promises)
  return str.replace(regex, () => replacements.shift()!)
}

let base: ResolvedConfig['base']
let assetsDir: ResolvedConfig['build']['assetsDir']
let outDir: ResolvedConfig['build']['outDir']
let logger: ResolvedConfig['logger']
let mode: ResolvedConfig['mode']
let root: ResolvedConfig['root']
let resolveVitePath: ResolveFn
let islandsConfig: IslandsConfig

const hydrationBegin = '<!--VITE_ISLAND_HYDRATION_BEGIN-->'
const hydrationEnd = '<!--VITE_ISLAND_HYDRATION_END-->'
const slotBegin = '<!--VITE_ISLAND_SLOT_BEGIN-->'
const slotSeparator = 'VITE_ISLAND_SLOT_SEPARATOR'
const commentsRegex = /<!--\[-->|<!--]-->/g
const hydrationRegex = new RegExp(`${escapeRegex(hydrationBegin)}(.*?)${escapeRegex(hydrationEnd)}`, 'sg')
const contextComponentRegex = new RegExp(escapeRegex('_ctx.__unplugin_components_'), 'g')
const unresolvedIslandKey = '__viteIslandComponent'
const viteIslandRegex = new RegExp(`"?${escapeRegex(unresolvedIslandKey)}"?:\\s*([^,]+),`, 'sg')
const scriptTagsRegex = /<script\s*([^>]*?)>(.*?)<\/script>/sg

const islandsByRoute: Record<string, string[]> = Object.create(null)

function config (config: UserConfig) {
  return {
    resolve: {
      alias: resolveAliases(config.root ?? process.cwd()),
    },
    server: {
      fs: {
        allow: [DIST_CLIENT_PATH, config.root ?? process.cwd()],
      },
    },
    build: {
      brotliSize: false,
      minify: false,
      ...config.build,
    },
    optimizeDeps: {
      include: [
        'vue',
        'vue-router',
        '@vueuse/core',
        '@vueuse/head',
        '@nuxt/devalue',
      ],
      exclude: [
        'vue-demi',
        'vue-islands',
      ],
    },
    islands: {
      tempDir: path.join(config.root || process.cwd(), '.vite-islands-temp'),
    },
    ssgOptions: {
      async onPageRendered (route, html) {
        let counter = 0
        const pageIslands: string[] = []
        const assetsBase = path.join(base, assetsDir)
        const pageOutDir = path.resolve(islandsConfig.tempDir, route === '/' ? 'index' : route.replace(/^\//, '').replace(/\//g, '-'))
        fs.mkdirSync(pageOutDir, { recursive: true })
        html = html.replace(scriptTagsRegex, (script: string, attrs: string) => {
          return !attrs.includes('client-keep') && attrs.includes('type="module"') && attrs.includes(`src="${assetsBase}`)
            ? ''
            : script
        })
        html = html.replace(/<link\s*([^>]*?)>/sg, (link: string, attrs: string) => {
          if (attrs.includes('modulepreload') && attrs.includes('.js')) return ''
          return link
        })
        html = await replaceAsync(html, hydrationRegex, async (str, slotsContent) => {
          const basename = `vite-island-${++counter}.js`
          const filename = path.join(pageOutDir, basename)
          const [scriptTemplate, ...slotStrs] = slotsContent.replace(commentsRegex, '').split(slotBegin)
          const slots = Object.fromEntries(slotStrs.map(str => str.split(slotSeparator)))
          const script = scriptTemplate.replace('/* VITE_ISLAND_HYDRATION_SLOTS */', slotStrs.length ? serialize(slots) : '')
          await fs.promises.writeFile(filename, script, 'utf-8')
          pageIslands.push(filename)
          return `${hydrationBegin}${filename}${hydrationEnd}`
        })
        if (pageIslands.length) {
          islandsByRoute[route] = pageIslands
          logger.warn(`${chalk.dim(`${path.relative(islandsConfig.tempDir, pageOutDir)}`)} ${chalk.green(` (${pageIslands.length})`)}`)
        }
        else {
          logger.warn(`${chalk.dim(`${path.relative(islandsConfig.tempDir, pageOutDir)}`)} ${chalk.yellow(' no islands')}`)
        }
        return html
      },
      async onFinished () {
        const islandFiles = Object.values(islandsByRoute).flat()
        if (islandFiles.length === 0) return

        buildLog('Build for islands...', islandFiles.length)

        const assetsBase = path.join(base, assetsDir)

        // Remove unnecessary client scripts.
        const files = await glob(path.join(outDir, '**/*.js'))
        files.forEach(fileName => fs.promises.rm(fileName))

        await viteBuild({
          publicDir: false,
          build: {
            minify: 'esbuild',
            brotliSize: true,
            emptyOutDir: false,
            manifest: true,
            outDir,
            rollupOptions: {
              input: islandFiles,
            },
          },
          mode,
        })
        const manifest: Manifest = JSON.parse(fs.readFileSync(path.join(outDir, 'manifest.json'), 'utf-8'))

        await Promise.all(Object.keys(islandsByRoute).map(async (route) => {
          const htmlFilename = filenameFromRoute(route)

          let html = await fs.promises.readFile(htmlFilename, 'utf-8')
          const preloadScripts: string[] = []
          html = await replaceAsync(html, hydrationRegex, async (str, file) => {
            const entry = manifest[path.relative(root, file)]
            if (entry.imports) preloadScripts.push(...entry.imports)

            const filename = path.join(outDir, entry.file)
            const code = await fs.promises.readFile(filename, 'utf-8')
            const rebasedCode = await rebaseImports(assetsBase, code)

            fs.promises.rm(filename)

            return `<script type="module">${rebasedCode}</script>`
          })
          html = html.replace('</head>', `${stringifyPreload(manifest, preloadScripts)}</head>`)
          await fs.promises.writeFile(htmlFilename, html, 'utf-8')
        }))
        fs.rmSync(islandsConfig.tempDir, { recursive: true, force: true })
      },
    } as ViteSSGOptions,
  }
}

// Public: Configures MDX, Vue, Components, and Islands plugins.
export default function IslandsPlugins (): PluginOption[] {
  return [
    {
      name: 'islands',
      resolveId (id) {
        if (id === SITE_DATA_REQUEST_PATH)
          return SITE_DATA_REQUEST_PATH

        if (id === ROUTES_REQUEST_PATH)
          return 'virtual:generated-pages'

        if (id === USER_CONFIG_REQUEST_PATH)
          return USER_CONFIG_REQUEST_PATH
      },
      load (id) {
        // TODO: Provide actual site data.
        const siteData = { base: '/' }

        if (id === SITE_DATA_REQUEST_PATH)
          return `export default ${JSON.stringify(JSON.stringify(siteData))}`

        if (id === USER_CONFIG_REQUEST_PATH) {
          const configPath = path.join(root, 'iles.config.ts')
          this.addWatchFile(configPath)
          return fs.readFileSync(configPath, 'utf-8')
        }
      },
      transform (code, id) {
        return code.replace(/__LAYOUTS_ROOT__/g, '/src/layouts')
      },
      configureServer (server) {
        // server.watcher.add(configPath)

        // serve our index.html after vite history fallback
        return () => {
          server.middlewares.use((req, res, next) => {
            if (req.url!.endsWith('.html')) {
              res.statusCode = 200
              res.end(`
  <!DOCTYPE html>
  <html>
    <head>
      <title></title>
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta name="description" content="">
    </head>
    <body>
      <div id="app"></div>
      <script type="module" src="/@fs/${APP_PATH}/index.js"></script>
    </body>
  </html>`)
              return
            }
            next()
          })
        }
      },
    },
    {
      name: 'islands:wrap-components',
      enforce: 'pre',
      config,
      configResolved (config) {
        debug.build('minify:', config.build.minify)
        if (base) return
        base = config.base
        assetsDir = config.build.assetsDir
        outDir = config.build.outDir
        logger = config.logger
        mode = config.mode
        root = config.root
        islandsConfig = (config as any).islands
        resolveVitePath = config.createResolver()
        debug.wrap({ outDir, assetsDir })
      },
      async transform (code, id) {
        const { path } = parseId(id)
        if (!path.endsWith('.mdx') && !path.endsWith('.vue')) return

        const components = /<([A-Z]\w+)\s*(?:([^/]+?)\/>|([^>]+)>(.*?)<\/\1>)/sg

        // Parse imports and set options as needed.
        const scriptContent = path.endsWith('.vue') && Array.from(code.matchAll(scriptTagsRegex)).map(([,, js]) => js).join(';')
        const imports = scriptContent
          ? await parseImports(scriptContent)
          : {}

        return code.replace(components, (str, tagName, attrs, otherAttrs, children) => {
          if (otherAttrs) attrs = otherAttrs
          if (!attrs?.match(/(\s|^)client:/)) return str

          debug.wrap(tagName, attrs)

          const resolveComponent = imports[tagName] ? tagName : `_resolveComponent("${tagName}")`
          const component = path.endsWith('.vue')
            ? `:${unresolvedIslandKey}='${resolveComponent}'`
            : `${unresolvedIslandKey}={${resolveComponent}}`

          return `<ViteIsland componentName="${pascalCase(tagName)}" ${component} ${attrs}>${children || ''}</ViteIsland>`
        })
      },
    },

    Vue({
      refTransform: true,
      template: {
        compilerOptions: {
          isCustomElement: (tagName: string) => tagName.startsWith('ile-'),
        },
      },
    }),

    XDM({
      jsx: true,
    }),

    {
      name: 'islands:mdx',
      async transform (code, id) {
        const { path } = parseId(id)
        if (!path.endsWith('.mdx') || !code.includes('MDXContent')) return null

        const match = code.match(/props\.components\), \{(.*?), wrapper: /)
        const importedComponents = match ? match[1].split(', ') : []
        debug.mdx(`transforming (${importedComponents.length})`, importedComponents)

        const pattern = '_components = Object.assign({'
        const index = code.indexOf(pattern) + pattern.length
        const comps = importedComponents.map(name => `    ${name}: _resolveComponent("${name}"),`).join('\n')
        code = `${code.slice(0, index)}\n${comps}\n${code.slice(index, code.length + 1)}`

        // Allow mdx pages with only frontmatter.
        code = code.replace('_content = <></>', '_content = null')

        // TODO: Allow component to receive an excerpt prop.
        return code.replace('export default MDXContent', `
          ${code.includes(' defineComponent') ? '' : 'import { defineComponent } from \'vue\''}

          const _default = defineComponent({
            ${mode === 'development' ? `__file: '${path}',` : ''}
            ...frontmatter,
            frontmatter,
            props: {
              components: { type: Object, default: () => ({}) },
            },
            render () {
              return MDXContent({ ...this.$props, ...this.$attrs })
            },
          })
          export const content = _default
          export default _default
        `)
      },
    },

    VueJSX({
      include: /\.[jt]sx|mdx?$/,
    }),

    // https://github.com/hannoeru/vite-plugin-pages
    Pages({
      extensions: ['vue', 'md', 'mdx'],
      extendRoute (route) {
        const file = path.resolve(root, route.component.slice(1))
        if (file.endsWith('.mdx') || file.endsWith('.md')) {
          const md = fs.readFileSync(file, 'utf-8')
          const { data: { layout, ...frontmatter } } = matter(md)
          route.meta = Object.assign(route.meta || {}, { frontmatter, layout })
          if (file.includes('posts/') && !file.endsWith('index.vue'))
            route.meta.layout = 'post'
        }
        return route
      },
    }),

    // https://github.com/JohnCampionJr/vite-plugin-vue-layouts
    Layouts(),

    // https://github.com/antfu/vite-plugin-components
    ViteComponents({
      dts: true,
      // allow auto load markdown components under `./src/components/`
      extensions: ['vue'],

      // allow auto import and register components used in markdown
      include: [/\.vue$/, /\.vue\?vue/, /\.mdx?/],

      // auto import icons
      resolvers: [
        // https://github.com/antfu/vite-plugin-icons
        IconsResolver({
          componentPrefix: '',
          // enabledCollections: ['carbon']
        }),
        (name) => {
          if (name === 'ViteIsland' || name === 'Island')
            return { importName: 'Island', path: 'iles' }
          else
            return null
        },
      ],
    }),

    {
      name: 'islands:import-analysis',
      enforce: 'post',
      async transform (code, id) {
        const { path } = parseId(id)
        if (!path.endsWith('.mdx') && !path.endsWith('.vue')) return

        code = code.replace(contextComponentRegex, '__unplugin_components_')

        if (!code.includes(unresolvedIslandKey)) return code

        const imports = await parseImports(code)
        code = await replaceAsync(code, viteIslandRegex, async (str, resolvedName) => {
          resolvedName = resolvedName.replace(/^\$setup\./, '')
          const importMetadata = imports[resolvedName]

          const componentName = resolvedName.match(/_resolveComponent\(([^)]+?)\)/)?.[1]

          if (!importMetadata)
            throw new Error(`Unable to infer '${componentName}' island component in ${path}`)

          debug.resolve(`${componentName} => ${importMetadata.path}`)

          return `
            component: ${resolvedName},
            importName: '${importMetadata.name}',
            importPath: '${await resolveVitePath(importMetadata.path, path)}',
          `
        })

        if (path.endsWith('.mdx')) {
          const name = code.match(/_resolveComponent\(([^)]+?)\)/)?.[1]
          if (name)
            throw new Error(`Unable to infer '${name}' component in ${path}`)
        }

        return code
      },
    },
  ]
}
