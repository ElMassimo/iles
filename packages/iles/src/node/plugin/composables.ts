import { promises as fs } from 'fs'
import { resolve } from 'pathe'
import { uniq } from './utils'
import { parseImports } from './parse'

const definitionRegex = /(?:function|const|let|var)\s+(definePageComponent|use(?:Page|Route|Head|Documents)\b)/g
const composableUsageRegex = /\b(definePageComponent|use(?:Page|Route|Head|Documents))\s*\(/g

export async function autoImportComposables (code: string, id: string): Promise<string | undefined> {
  const matches = Array.from(code.matchAll(composableUsageRegex))
  if (matches.length === 0) return

  const imports = await parseImports(code)
  const defined = new Set(Array.from(code.matchAll(definitionRegex)).map(a => a[1]))

  const composables = uniq(matches.map(a => a[1]))
    .filter(composable => !defined.has(composable) && !imports[composable])
    .join(', ')

  if (composables)
    return `${code}\nimport { ${composables} } from "iles"`
}

export function writeComposablesDTS (root: string) {
  fs.writeFile(resolve(root, 'composables.d.ts'), `// generated by iles
// We suggest you to commit this file into source control

declare global {
  const definePageComponent: typeof import('iles')['definePageComponent']
  const useHead: typeof import('iles')['useHead']
  const usePage: typeof import('iles')['usePage']
  const useRoute: typeof import('iles')['useRoute']
}

export { }
`, 'utf-8')
}
