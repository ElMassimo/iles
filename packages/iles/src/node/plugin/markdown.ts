import type { ViteDevServer } from 'vite'
import deepEqual from 'deep-equal'
import type { AppConfig } from '../shared'

let originalTags: string[]

// Internal: Detects markdown components overriden in the app.
export function detectMDXComponents (code: string, config: AppConfig, server?: ViteDevServer | undefined) {
  const mdxComponents = code.match(/\bmdxComponents\b(?:.*?){(.*?)}/s)?.[1]
  if (!mdxComponents) return

  const foundTags = Array.from(mdxComponents.matchAll(/\b['"]?(\w+)['"]?:/g)).map(m => m[1])

  if (!originalTags)
    originalTags = config.markdown.overrideElements ||= []

  const dynamicElements = Array.from(new Set([...originalTags, ...foundTags])).sort()
  if (!deepEqual(dynamicElements, config.markdown.overrideElements)) {
    config.markdown.overrideElements = dynamicElements
    server?.moduleGraph.invalidateAll()
  }
}
