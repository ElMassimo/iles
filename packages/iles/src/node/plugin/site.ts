import type { AppConfig } from '../shared'

// Internal: Adds the url to the site for convenience, and enables HMR.
export function extendSite(code: string, config: AppConfig) {
  return `${code.replace('export default ', 'let __site = ')}
__site.url = '${config.siteUrl}${config.base.slice(0, config.base.length - 1)}'
__site.canonical = '${config.siteUrl.split('//', 2)[1] ?? ''}'
import { ref as _$ref } from 'vue'
const __siteRef = _$ref(__site)
__site = { ref: __siteRef  }
export { __site, __siteRef as default }

if (import.meta.hot)
  import.meta.hot.accept(mod => {
    __site.ref.value = mod.__site.ref.value
    mod.__site.ref = __site.ref
  })
`
  // NOTE: The last line replaces the ref of the current module with the ref in
  // the original module that was made reactive in `installPageData`, so that
  // subsequent HMRs are also performed as expected.
}
