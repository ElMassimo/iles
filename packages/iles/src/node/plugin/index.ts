/* eslint-disable no-restricted-syntax */
import { promises as fs } from 'fs'
import { basename, resolve, relative } from 'pathe'
import { green } from 'nanocolors'
import type { PluginOption, ResolvedConfig, ResolveFn, ViteDevServer } from 'vite'
import { transformWithEsbuild } from 'vite'

import { MODULE_ID_VIRTUAL as PAGES_REQUEST_PATH } from 'vite-plugin-pages'
import MagicString from 'magic-string'

import type { Frontmatter } from '@islands/frontmatter'
import createDebugger from 'debug'
import type { AppConfig, AppClientConfig } from '../shared'
import { APP_PATH, ROUTES_REQUEST_PATH, USER_APP_REQUEST_PATH, USER_SITE_REQUEST_PATH, APP_CONFIG_REQUEST_PATH, NOT_FOUND_COMPONENT_PATH, NOT_FOUND_REQUEST_PATH } from '../alias'
import { createServer } from '../server'
import { escapeRegex, serialize, replaceAsync, pascalCase, exists } from './utils'
import { parseId, parseImports } from './parse'
import { unresolvedIslandKey, wrapIslandsInSFC, wrapLayout } from './wrap'
import { extendSite } from './site'

const debug = {
  config: createDebugger('iles:config'),
  mdx: createDebugger('iles:mdx'),
  layout: createDebugger('iles:layout'),
  detect: createDebugger('iles:detect'),
  resolve: createDebugger('iles:resolve'),
  build: createDebugger('iles:build'),
}

function isMarkdown (path: string) {
  return path.endsWith('.mdx') || path.endsWith('.md')
}

function isSFCMain (path: string, query: Record<string, any>) {
  return path.endsWith('.vue') && query.vue === undefined
}

const contextComponentRegex = new RegExp(escapeRegex('_ctx.__unplugin_components_'), 'g')
const viteIslandRegex = new RegExp(`"?${escapeRegex(unresolvedIslandKey)}"?:\\s*([^,}\n]+)[,}\n]`, 'sg')
const unresolvedComponentsRegex = /_resolveComponent\("([^)]+?)"\)/
const templateLayoutRegex = /<template.*?\slayout=\s*['"](\w+)['"].*?>/

// Public: Configures MDX, Vue, Components, and Islands plugins.
export default function IslandsPlugins (appConfig: AppConfig): PluginOption[] {
  debug.config(appConfig)

  let base: ResolvedConfig['base']
  let mode: ResolvedConfig['mode']
  let root: ResolvedConfig['root']
  let resolveVitePath: ResolveFn

  const appPath = resolve(appConfig.srcDir, 'app.ts')
  const sitePath = resolve(appConfig.srcDir, 'site.ts')
  const layoutsRoot = `/${relative(appConfig.root, appConfig.layoutsDir)}`
  const defaultLayoutPath = `${layoutsRoot}/default.vue`

  const plugins = appConfig.namedPlugins

  function isLayout (path: string) {
    return path.includes(appConfig.layoutsDir)
  }

  function frontmatterFromPage (path: string): Frontmatter | undefined {
    if (plugins.pages.api.pageForFile(path))
      return appConfig.markdown.extendFrontmatter?.({}, path) || {}
  }

  return [
    {
      name: 'iles',
      enforce: 'pre',
      configResolved (config) {
        if (base) return
        base = config.base
        mode = config.mode
        root = config.root
        resolveVitePath = config.createResolver()
      },
      async resolveId (id) {
        if (id === ROUTES_REQUEST_PATH)
          return PAGES_REQUEST_PATH

        if (id === APP_CONFIG_REQUEST_PATH || id === USER_APP_REQUEST_PATH || id === USER_SITE_REQUEST_PATH)
          return id

        if (id === NOT_FOUND_REQUEST_PATH) {
          for (const extension of ['vue', 'mdx', 'md', 'jsx']) {
            const path = resolve(appConfig.pagesDir, `404.${extension}`)
            if (await exists(path)) return path
          }
          return NOT_FOUND_COMPONENT_PATH
        }

        // Prevent import analysis failure if the default layout doesn't exist.
        if (id === defaultLayoutPath) return resolve(root, id.slice(1))
      },
      async load (id) {
        if (id === APP_CONFIG_REQUEST_PATH) {
          const { base, debug, jsx, root, ssg: { manualChunks: _, ...ssg }, siteUrl } = appConfig
          const clientConfig: AppClientConfig = { base, debug, root, jsx, ssg, siteUrl }
          return `export default ${serialize(clientConfig)}`
        }

        const userFilename = (id === USER_APP_REQUEST_PATH && appPath)
          || (id === USER_SITE_REQUEST_PATH && sitePath)
        if (userFilename) {
          this.addWatchFile(userFilename)
          const result = await exists(userFilename)
            ? await transformWithEsbuild(await fs.readFile(userFilename, 'utf-8'), userFilename, { sourcemap: false })
            : { code: 'export default {}' }
          return id === USER_SITE_REQUEST_PATH ? extendSite(result.code, appConfig) : result
        }
      },
      handleHotUpdate ({ file, server }) {
        if (file === appPath) return [server.moduleGraph.getModuleById(USER_APP_REQUEST_PATH)!]
        if (file === sitePath) return [server.moduleGraph.getModuleById(USER_SITE_REQUEST_PATH)!]
      },
      configureServer (server) {
        restartOnConfigChanges(appConfig, server)

        const supportedExtensions = ['.html', '.xml', '.json', '.rss', '.atom']

        // serve our index.html after vite history fallback
        return () => {
          server.middlewares.use((req, res, next) => {
            // Fallback when the user has not created a default layout.
            if (req.url?.includes(defaultLayoutPath)) {
              res.statusCode = 200
              res.setHeader('content-type', 'text/javascript')
              res.end('export default false')
            }
            else if (supportedExtensions.some(ext => req.url!.endsWith(ext))) {
              res.statusCode = 200
              res.setHeader('content-type', 'text/html')
              res.end(`
<!DOCTYPE html>
<html>
  <body>
    <div id="app"></div>
    <script type="module" src="/@fs/${APP_PATH}/index.js"></script>
  </body>
</html>`)
            }
          })
        }
      },
    },
    {
      name: 'iles:detect-islands-in-vue',
      enforce: 'pre',
      async transform (code, id) {
        const { path, query } = parseId(id)

        if (query.vue !== undefined && query.type === 'scriptClient')
          return 'export default {}; if (import.meta.hot) import.meta.hot.accept()'

        if (isSFCMain(path, query) && code.includes('client:') && code.includes('<template'))
          return wrapIslandsInSFC(code, path, debug.detect)
      },
    },
    {
      name: 'iles:layouts',
      enforce: 'pre',
      transform (code, id) {
        const { path, query } = parseId(id)
        if (!isSFCMain(path, query) || !isLayout(path)) return
        const layoutName = code.match(templateLayoutRegex)?.[1] || false
        if (String(layoutName) === 'false') return
        return wrapLayout(code, path, debug.layout)
      },
    },

    plugins.vue,
    ...plugins.optionalPlugins,

    plugins.markdown,

    {
      name: 'iles:mdx:pos',
      async transform (code, id) {
        const { path } = parseId(id)
        if (!isMarkdown(path) || !code.includes('MDXContent')) return null

        const match = code.match(/props\.components\), \{(.*?), wrapper: /)
        const importedComponents = match ? match[1].split(', ') : []
        debug.mdx(`transforming (${importedComponents.length})`, importedComponents)

        const pattern = '_components = Object.assign({'
        const index = code.indexOf(pattern) + pattern.length
        const comps = importedComponents.map(name => `    ${name}: _resolveComponent("${name}"),`).join('\n')
        code = `${code.slice(0, index)}\n${comps}\n${code.slice(index, code.length + 1)}`

        // Allow mdx pages with only frontmatter.
        code = code.replace('_content = <></>', '_content = null')

        // Allow MDX content to be rendered without args.
        code = code.replace('MDXContent(props)', 'MDXContent(props = {})')

        // TODO: Allow component to receive an excerpt prop.
        return code.replace('export default MDXContent', `
${code.includes(' defineComponent') ? '' : 'import { defineComponent } from \'vue\''}

const _sfc_main = defineComponent({
  ${mode === 'development' ? `__file: '${path}',` : ''}
  props: {
    components: { type: Object, default: () => ({}) },
  },
  render () {
    return MDXContent({ ...this.$props, ...this.$attrs })
  },
})
export const render = MDXContent
export default _sfc_main
`)
      },
    },

    plugins.vueJsx,

    // https://github.com/hannoeru/vite-plugin-pages
    plugins.pages,

    // https://github.com/antfu/unplugin-vue-components
    plugins.components,

    {
      name: 'iles:page-hmr',
      apply: 'serve',
      enforce: 'post',
      // Force a refresh for all page computed properties.
      async transform (code, id) {
        const { path } = parseId(id)
        if (isLayout(path) || plugins.pages.api.pageForFile(path)) {
          return `${code}
import.meta.hot.accept('/${relative(root, path)}', (...args) => __ILES_PAGE_UPDATE__(args))
`
        }
      },
    },

    {
      name: 'iles:sfc:page-data',
      enforce: 'post',
      async transform (code, id) {
        const { path, query } = parseId(id)
        const isMarkdownPath = isMarkdown(path)
        if (!isMarkdownPath && !isSFCMain(path, query)) return

        const s = new MagicString(code)
        const sfcIndex = code.indexOf('{', code.indexOf('const _sfc_main = ')) + 1
        const appendToSfc = (key: string, value: string) => s.appendRight(sfcIndex, `${key}:${value},`)

        if (isLayout(path)) {
          appendToSfc('name', `'${pascalCase(basename(path).replace('.vue', 'Layout'))}'`)
          return s.toString()
        }

        const pageMatter = frontmatterFromPage(path)
        if (!pageMatter) return

        if (isMarkdownPath) {
          s.appendRight(sfcIndex, '...meta,...frontmatter,meta,frontmatter,')
        }
        else {
          const { meta, layout, ...frontmatter } = pageMatter
          appendToSfc('meta', serialize(meta))
          appendToSfc('frontmatter', serialize(frontmatter))
        }

        const layoutName = pageMatter.layout ?? 'default'
        appendToSfc('layoutName', serialize(layoutName))
        appendToSfc('layoutFn', String(layoutName) === 'false'
          ? 'false'
          : `() => import('${layoutsRoot}/${layoutName}.vue').then(m => m.default)`)

        return s.toString()
      },
    },

    {
      name: 'iles:client-scripts',
      enforce: 'post',
      async transform (code, id) {
        const { path, query } = parseId(id)

        if (query.clientScript)
          return await transformWithEsbuild(code, path, { loader: query.lang })
      },
    },

    {
      name: 'iles:resolve-islands',
      enforce: 'post',
      async transform (code, id) {
        const { path, query } = parseId(id)
        if (!isMarkdown(path) && !isSFCMain(path, query)) return

        code = code.replace(contextComponentRegex, '__unplugin_components_')

        if (!code.includes(unresolvedIslandKey)) return code

        const imports = await parseImports(code)

        code = await replaceAsync(code, viteIslandRegex, async (str, resolvedName) => {
          resolvedName = resolvedName.replace(/^\$setup\./, '')
          const componentName = resolvedName.match(unresolvedComponentsRegex)?.[1] || resolvedName
          const importMetadata = imports[componentName]

          if (!importMetadata)
            throw new Error(`Unable to infer '${componentName}' island component in ${path}`)

          debug.resolve(`${componentName} => ${importMetadata.path}`)

          return `
component: ${componentName},
importName: '${importMetadata.name}',
importPath: '${await resolveVitePath(importMetadata.path, path)}',
          `.replace(/\n/g, ' ')
        })

        if (isMarkdown(path)) {
          const componentName = code.match(unresolvedComponentsRegex)?.[1]
          if (componentName)
            throw new Error(`Unable to infer '${componentName}' component in ${path}`)
        }

        return code
      },
    },
  ]
}

async function restartOnConfigChanges (config: AppConfig, server: ViteDevServer) {
  const restartIfConfigChanged = async (path: string) => {
    if (path === config.configPath) {
      server.config.logger.info(
        green(
          `${relative(process.cwd(), config.configPath)} changed, restarting server...`,
        ),
        { clear: true, timestamp: true },
      )
      await server.close()
      // @ts-ignore
      global.__vite_start_time = Date.now()
      const { server: newServer } = await createServer(server.config.root, server.config.server)
      await newServer.listen()
    }
  }
  // Shut down the server and start a new one if config changes.
  server.watcher.add(config.configPath)
  server.watcher.on('add', restartIfConfigChanged)
  server.watcher.on('change', restartIfConfigChanged)
}
