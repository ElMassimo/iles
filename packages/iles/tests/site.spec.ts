import { test, describe, expect, beforeAll } from 'vitest'

import { extendSite } from '@node/plugin/site'

describe('site', () => {
  test('without data', () => {
    const extended = extendSite(`
export default {}
`, { siteUrl: 'http://example.com', base: '/' })
    expect(extended).toMatchSnapshot()
  })

  test('without site url', () => {
    const extended = extendSite(`
const site = {
  title: 'îles',
}

export default site
`, { siteUrl: '', base: '/' })
    expect(extended).toMatchSnapshot()
  })

  test('with site url', () => {
    const extended = extendSite(`
export default {
  title: 'îles',
}
`, { siteUrl: 'https://example.com', base: '/awesome/' })
    expect(extended).toMatchSnapshot()
  })
})
