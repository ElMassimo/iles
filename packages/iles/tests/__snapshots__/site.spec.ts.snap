// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`site > with site url 1`] = `
"
let __site = {
  title: 'îles',
}

__site.url = 'https://example.com/awesome'
__site.canonical = 'example.com'
import { ref as _$ref } from 'vue'
const __siteRef = _$ref(__site)
__site = { ref: __siteRef  }
export { __site, __siteRef as default }

if (import.meta.hot)
  import.meta.hot.accept(mod => {
    __site.ref.value = mod.__site.ref.value
    mod.__site.ref = __site.ref
  })
"
`;

exports[`site > without data 1`] = `
"
let __site = {}

__site.url = 'http://example.com'
__site.canonical = 'example.com'
import { ref as _$ref } from 'vue'
const __siteRef = _$ref(__site)
__site = { ref: __siteRef  }
export { __site, __siteRef as default }

if (import.meta.hot)
  import.meta.hot.accept(mod => {
    __site.ref.value = mod.__site.ref.value
    mod.__site.ref = __site.ref
  })
"
`;

exports[`site > without site url 1`] = `
"
const site = {
  title: 'îles',
}

let __site = site

__site.url = ''
__site.canonical = ''
import { ref as _$ref } from 'vue'
const __siteRef = _$ref(__site)
__site = { ref: __siteRef  }
export { __site, __siteRef as default }

if (import.meta.hot)
  import.meta.hot.accept(mod => {
    __site.ref.value = mod.__site.ref.value
    mod.__site.ref = __site.ref
  })
"
`;
